Add __GIT_REV__ and __GIT_SHA__ variables for use at build time

deldrid1 opened the pull request:
I haven't done extensive testing on this, but it meets my goal of having better version information embeddable into my Squirrel.

Could you review, and let me know your thoughts on the PR?  (P.S., I haven't updated any of the version information but I assume this is MINOR bump)

ppetrosh commented on src/index.js:
```diff
@@ -99,6 +100,30 @@ class Builder {
+      // Set the git commit sha
+      this._globals["__GIT_SHA__"] = execSync("git rev-parse HEAD").toString().trimRight()
+
+    } catch(e){}  // Silently fail on error - git is not in the path or we aren't in a git repo
```
I would still print a diagnostic error so that it's clear what's going on (in both cases).

ppetrosh commented:
I would probably request the values only if one uses them in the code.

deldrid1 commented:
I’m not super familiar with Builders code structure.... where would I put something like that?

ppetrosh commented:
Let us look into it.

ppetrosh commented:
I'm leaning towards maybe implementing this as a library function https://github.com/electricimp/builder#including-javascript-libraries. They are going to be called only at the moment of execution and will be project specific. You don't have to modify the builder itself to access them. And we would publish them as examples.

What do you think @deldrid1 ?

deldrid1 commented:
Yea that works for me - I didn’t think about implementing them there. 

That would give people more control over them as well if they want to modify the behavior in some way.

ppetrosh commented:
Exactly, as I start feeling that the format that you've proposed might be too application specific and I can imagine people wanting something else...

deldrid1 commented:
@ppetrosh - I think you can close this PR.  I've now implemented a JS `gitMethods.js` lib.

```javascript
const execSync = require('child_process').execSync

module.exports = {
    GitRepoDirty: () =>{
        try{

            var gitStatus = execSync('git status')

            if(gitStatus.includes("modified:") || gitStatus.includes("Untracked files:") || gitStatus.includes("deleted:")){
               return true
            }

           return false

        } catch(e){
            console.error("Unable to determine GitRev - are you in a git repo with 'git' in your PATH?")
            console.error(e)
            throw e
        }
    },

    GitRevInt: () => {
        try{
            //Parse output from git describe and turn it into a 32-bit bit field.
            //Bits 0 – 27: These encode a 7-digit Git revision ID of the firmware build. This is equivalent to the first 7 characters of “git rev-parse --short HEAD”. For example, if this command reports a revision ID of "a4d5ca3c", this field will be 0xa4d5ca3.
            //Bits 28 – 31: The top 4 bits are flags.
                // Bit 31 = "dirty".            Used to indicate there were uncommitted changes when the binary was built.
                // Bit 30 = "File modified".    Used to indicate modified files are present
                // Bit 29 = "File Added".       Used to indicate that a file has been added
                // Bit 28 = "File Removed".     Used to indicate that a file has been deleted

            var int32Val = parseInt("0x" + execSync('git rev-parse --short HEAD')) >>> 4

            var gitStatus = execSync('git status')

            if(gitStatus.includes("modified:")){
                int32Val |= (1 << 31)
                int32Val |= (1 << 30)
            }

            if(gitStatus.includes("Untracked files:")){
                int32Val |= (1 << 31)
                int32Val |= (1 << 29)
            }

            if(gitStatus.includes("deleted:")){
                int32Val |= (1 << 31)
                int32Val |= (1 << 28)
            }

            return "0x" + (int32Val >>> 0).toString(16)

        } catch(e){
            console.error("Unable to determine GitRev - are you in a git repo with 'git' in your PATH?")
            console.error(e)
        }
    },

    GitSha: () => {
        try{
            return execSync("git rev-parse HEAD").toString().trimRight()
        } catch(e){
            console.error("Unable to determine GitSha - are you in a git repo with 'git' in your PATH?")
            console.error(e)
        }
    },

    GitUserName: () => {
        try{
            return execSync('git config user.name').toString().slice(0, -1) //Get rid of trailing new line
        } catch(e){
            console.error("Unable to retrieve your git User Name - is 'git' on your PATH?")
            console.error(e)
            throw e
        }
    },

    GitOriginURL: (idx=0) => {
        //TODO: This is an imprecise science - we need to make a more intelligent parser here...
        // Take output like
        // origin	https://github.com/deldrid1/Builder.git (fetch)
        // origin	https://github.com/deldrid1/Builder.git (push)
        //
        // And turn it into output like
        // https://github.com/deldrid1/Builder

        try{
            console.log(execSync('git remote -v').toString()
            .split("\n")[idx]   // Pull out the origin we want by new line
            .split("\t")[1]     // Extract starting at https://
            .split(" ")[0].slice(0, -4)) // Pull out the url, minus ".git")

            return execSync('git remote -v').toString()
                    .split("\n")[idx]   // Pull out the origin we want by new line
                    .split("\t")[1]     // Extract starting at https://
                    .split(" ")[0].toString().slice(0, -4) // Pull out the url, minus ".git"
        } catch(e){
            console.error("Unable to determine GitOriginURL - are you in a git repo with 'git' in your PATH?")
            console.error(e)
            throw e
        }
    }
}

```

## Prebuilt code

```squirrel
FW_VERSION <- @{GitRevInt()} 
const FW_GIT_SHA = "@{GitSha()}"

server.log("Running build prepared by git user '@{GitUserName()}' with git SHA '" + FW_GIT_SHA + "'")
server.log("Repo origin = @{GitOriginURL()}/commit/" + FW_GIT_SHA)
//FW_VERSION is set by the Builder GitRevInt() User defined JS method.
//Bits 0 – 27: encode a 7-digit Git revision ID of the firmware build. This is equivalent to the first 7 characters of “git rev-parse --short HEAD”. For example, if this command reports a revision ID of "a4d5ca3c", this field will be 0xa4d5ca3.
//Bits 28 – 31: The top 4 bits are flags.
    // Bit 31 = "Repo Dirty".            Used to indicate there were uncommitted changes when the binary was built.
    // Bit 30 = "File modified".    Used to indicate modified files are present
    // Bit 29 = "File Added".       Used to indicate that a file has been added
    // Bit 28 = "File Removed".     Used to indicate that a file has been deleted
if((FW_VERSION & 0x80000000) == 0x80000000){
    server.log("[WARNING]: There were uncommitted changes when the binary was built.")

    if((FW_VERSION & 0x40000000) == 0x40000000)
        server.log("\t[WARNING]: Tracked Files have been modified since '" + FW_GIT_SHA + "'")

    if((FW_VERSION & 0x20000000) == 0x20000000)
        server.log("\t[WARNING]: Files have been added to the repo since '" + FW_GIT_SHA + "'")

    if((FW_VERSION & 0x10000000) == 0x10000000)
        server.log("\t[WARNING]: Tracked Files have been removed from the repo since '" + FW_GIT_SHA + "'")

}

local currentTime = date(time(), 'u');
server.log("[" + __EI.PRODUCT_NAME + "][" + __EI.DEVICEGROUP_NAME + "]: " + imp.configparams.deviceid + " Agent Booted and Squirrel Init Complete - " + format("%.4d-%.2d-%.2dT%.2d:%.2d:%.2d (%d)", currentTime.year, currentTime.month+1, currentTime.day,  currentTime.hour, currentTime.min, currentTime.sec, time()))
server.log(format("'%s' Device Group ID '%s' with Deployment ID '%s'",__EI.DEVICEGROUP_TYPE, __EI.DEVICEGROUP_ID, __EI.DEPLOYMENT_ID))
server.log(format("Code SHA '%s' created at '%s'", __EI.DEPLOYMENT_SHA, __EI.DEPLOYMENT_CREATED_AT));
server.log(format("Running OS Version '%s', with Free Memory %d", imp.getsoftwareversion(), imp.getmemoryfree()))

```

## Built code
```squirrel
FW_VERSION <- 0xea4d5ca3
const FW_GIT_SHA = "a4d5ca3cfdd4f6d363ceb6530ef0fad80217108f"

server.log("Running build prepared by git user 'Austin Eldridge' with git SHA '" + FW_GIT_SHA + "'")
server.log("Repo origin = https://deldrid1@github.com/deldrid1/my-private-repo/commit/" + FW_GIT_SHA)
//FW_VERSION is set by the Builder GitRevInt() User defined JS method.
//Bits 0 – 27: encode a 7-digit Git revision ID of the firmware build. This is equivalent to the first 7 characters of “git rev-parse --short HEAD”. For example, if this command reports a revision ID of "a4d5ca3c", this field will be 0xa4d5ca3.
//Bits 28 – 31: The top 4 bits are flags.
    // Bit 31 = "Repo Dirty".            Used to indicate there were uncommitted changes when the binary was built.
    // Bit 30 = "File modified".    Used to indicate modified files are present
    // Bit 29 = "File Added".       Used to indicate that a file has been added
    // Bit 28 = "File Removed".     Used to indicate that a file has been deleted
if((FW_VERSION & 0x80000000) == 0x80000000){
    server.log("[WARNING]: There were uncommitted changes when the binary was built.")

    if((FW_VERSION & 0x40000000) == 0x40000000)
        server.log("\t[WARNING]: Tracked Files have been modified since '" + FW_GIT_SHA + "'")

    if((FW_VERSION & 0x20000000) == 0x20000000)
        server.log("\t[WARNING]: Files have been added to the repo since '" + FW_GIT_SHA + "'")

    if((FW_VERSION & 0x10000000) == 0x10000000)
        server.log("\t[WARNING]: Tracked Files have been removed from the repo since '" + FW_GIT_SHA + "'")

}

local currentTime = date(time(), 'u');
server.log("[" + __EI.PRODUCT_NAME + "][" + __EI.DEVICEGROUP_NAME + "]: " + imp.configparams.deviceid + " Agent Booted and Squirrel Init Complete - " + format("%.4d-%.2d-%.2dT%.2d:%.2d:%.2d (%d)", currentTime.year, currentTime.month+1, currentTime.day,  currentTime.hour, currentTime.min, currentTime.sec, time()))
server.log(format("'%s' Device Group ID '%s' with Deployment ID '%s'",__EI.DEVICEGROUP_TYPE, __EI.DEVICEGROUP_ID, __EI.DEPLOYMENT_ID))
server.log(format("Code SHA '%s' created at '%s'", __EI.DEPLOYMENT_SHA, __EI.DEPLOYMENT_CREATED_AT));
server.log(format("Running OS Version '%s', with Free Memory %d", imp.getsoftwareversion(), imp.getmemoryfree()))
```

## Logs!
```
2019-02-21T16:59:35 +00:00	[Agent]	Running build prepared by git user 'Austin Eldridge' with git SHA 'a4d5ca3cfdd4f6d363ceb6530ef0fad80217108f'
2019-02-21T16:59:35 +00:00	[Agent]	Repo origin = https://deldrid1@github.com/deldrid1/my-private-repo/commit/a4d5ca3cfdd4f6d363ceb6530ef0fad80217108f
2019-02-21T16:59:35 +00:00	[Agent]	[WARNING]: There were uncommitted changes when the binary was built.
2019-02-21T16:59:35 +00:00	[Agent]		[WARNING]: Tracked Files have been modified since 'a4d5ca3cfdd4f6d363ceb6530ef0fad80217108f'
2019-02-21T16:59:35 +00:00	[Agent]		[WARNING]: Files have been added to the repo since 'a4d5ca3cfdd4f6d363ceb6530ef0fad80217108f'
2019-02-21T16:59:35 +00:00	[Agent]	[__GHOSTS][E0088611 (AE) Development]: 20000c2a690a3756 Agent Booted and Squirrel Init Complete - 2019-02-21T16:59:35 (1550768375)
2019-02-21T16:59:35 +00:00	[Agent]	'development' Device Group ID '1625cd45-328e-3ffe-231d-b3b98f4171ad' with Deployment ID '3a4b6cf0-7b6e-7905-5756-fd15fd0b1332'
2019-02-21T16:59:35 +00:00	[Agent]	Code SHA '53b02031f8b3430e0a3fe981116fefff9b60f2b754b7f4a0b945f5e19e65d2f9' created at '2019-02-21T16:59:33.866Z'
2019-02-21T16:59:35 +00:00	[Agent]	Running OS Version '5ffc789 - jenkins-agent-runtime-release-14.04-31 - Mon Jan 7 12:48:43 2019 - Electric Imp', with Free Memory 1025751
```

deldrid1 commented:
Feel free to incorporate this example anywhere you like if it helps someone else!

ppetrosh commented:
Thank you @deldrid1!
